apis/kubedb/v1alpha2/mssql_webhook.go

```diff
@@ -0,0 +1,286 @@
+		Complete()
+}
+
+//+kubebuilder:webhook:path=/mutate-kubedb-com-v1alpha2-mssql,mutating=true,failurePolicy=fail,sideEffects=None,groups=kubedb.com,resources=mssqls,verbs=create;update,versions=v1alpha2,name=mmssql.kb.io,admissionReviewVersions=v1
+
+var _ webhook.Defaulter = &MsSQL{}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type
+func (m *MsSQL) Default() {
+	if m == nil {
+		return
+	}
+	mssqllog.Info("default", "name", m.Name)
+
+	m.SetDefaults()
+}
+
+//+kubebuilder:webhook:path=/validate-kubedb-com-v1alpha2-mssql,mutating=false,failurePolicy=fail,sideEffects=None,groups=kubedb.com,resources=mssqls,verbs=create;update,versions=v1alpha2,name=vmssql.kb.io,admissionReviewVersions=v1
+
+var _ webhook.Validator = &MsSQL{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (m *MsSQL) ValidateCreate() (admission.Warnings, error) {
+	mssqllog.Info("validate create", "name", m.Name)
+
+	allErr := m.ValidateCreateOrUpdate()
+	if len(allErr) == 0 {
+		return nil, nil
+	}
+	return nil, apierrors.NewInvalid(schema.GroupKind{Group: "kubedb.com", Kind: "MsSQL"}, m.Name, allErr)
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (m *MsSQL) ValidateUpdate(old runtime.Object) (admission.Warnings, error) {
+	mssqllog.Info("validate update", "name", m.Name)
+
+	oldMsSQL := old.(*MsSQL)
+	allErr := m.ValidateCreateOrUpdate()
+
+	if m.Spec.Topology == nil && *oldMsSQL.Spec.Replicas == 1 && *m.Spec.Replicas > 1 {
+		allErr = append(allErr, field.Invalid(field.NewPath("spec").Child("replicas"),
+			m.Name,
+			"Cannot scale up from 1 to more than 1 in standalone mode"))
+	}
+
+	if len(allErr) == 0 {
+		return nil, nil
+	}
+
+	return nil, apierrors.NewInvalid(schema.GroupKind{Group: "kubedb.com", Kind: "MsSQL"}, m.Name, allErr)
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
+func (m *MsSQL) ValidateDelete() (admission.Warnings, error) {
+	mssqllog.Info("validate delete", "name", m.Name)
+
+	var allErr field.ErrorList
+	if m.Spec.TerminationPolicy == TerminationPolicyDoNotTerminate {
+		allErr = append(allErr, field.Invalid(field.NewPath("spec").Child("terminationPolicy"),
+			m.Name,
+			"Can not delete as terminationPolicy is set to \"DoNotTerminate\""))
+		return nil, apierrors.NewInvalid(schema.GroupKind{Group: "kubedb.com", Kind: "MsSQL"}, m.Name, allErr)
+	}
+	return nil, nil
+}
+
+func (m *MsSQL) ValidateCreateOrUpdate() field.ErrorList {
+	var allErr field.ErrorList
+
+	if m.Spec.Version == "" {
+		allErr = append(allErr, field.Invalid(field.NewPath("spec").Child("version"),
+			m.Name,
+			"spec.version' is missing"))
+	} else {
+		err := mssqlValidateVersion(m)
+		if err != nil {
+			allErr = append(allErr, field.Invalid(field.NewPath("spec").Child("version"),
+				m.Name,
+				err.Error()))
+		}
+	}
+
+	if m.IsStandalone() {
+		if *m.Spec.Replicas != 1 {
```

Comment:
Give a nil check and then use "k8s.io/utils/ptr" ptr.Deref

Reply:
If we use ptr.Deref(), why do we need the nil check? 
ptr.Deref(m.Spec.Replicas, 0) != 1 check is fine I guess.